pkg/controller/servicemesh/controlplane/reconciler.go

```diff
@@ -383,27 +411,36 @@ func (r *ControlPlaneReconciler) renderCharts() error {
 }
 
 func (r *ControlPlaneReconciler) PostStatus() error {
-	return r.Client.Status().Update(context.TODO(), r.Instance)
+	instance := &v1.ServiceMeshControlPlane{}
+	if updateErr := r.Client.Get(context.TODO(), client.ObjectKey{Name: r.Instance.Name, Namespace: r.Instance.Namespace}, instance); updateErr == nil {
+		instance.Status = *r.Status.DeepCopy()
+		if updateErr = r.Client.Status().Update(context.TODO(), instance); updateErr != nil && !(apierrors.IsGone(updateErr) || apierrors.IsNotFound(updateErr)) {
+			r.Log.Error(updateErr, "error updating ServiceMeshControlPlane status")
+			return updateErr
+		}
+	} else if !(apierrors.IsGone(updateErr) || apierrors.IsNotFound(updateErr)) {
+		r.Log.Error(updateErr, "error updating ServiceMeshControlPlane status")
+		return updateErr
+	}
+	return nil
 }
 
 func (r *ControlPlaneReconciler) postReconciliationStatus(reconciliationMessage string, processingErr error) {
 	reconciledCondition := r.Status.GetCondition(v1.ConditionTypeReconciled)
 	if processingErr == nil {
 		reconciledCondition.Message = reconciliationMessage
 	} else {
-		reconciledCondition.Message = fmt.Sprintf("reconciliation error: %s", processingErr)
-	}
-	r.Status.SetCondition(reconciledCondition)
-	instance := &v1.ServiceMeshControlPlane{}
-	if updateErr := r.Client.Get(context.TODO(), client.ObjectKey{Name: r.Instance.Name, Namespace: r.Instance.Namespace}, instance); updateErr == nil {
-		instance.Status = *r.Status
-		r.Instance = instance
-		if updateErr = r.PostStatus(); updateErr != nil && !(apierrors.IsGone(updateErr) || apierrors.IsNotFound(updateErr)) {
-			r.Log.Error(updateErr, "error updating ServiceMeshControlPlane status")
+		reconciledCondition.Message = fmt.Sprintf("%s: error: %s", reconciliationMessage, processingErr)
+		var reason string
+		if r.Status.ObservedGeneration == 0 {
+			reason = "CreatingServiceMesh"
+		} else {
+			reason = "UpdatingServiceMesh"
 		}
-	} else if !(apierrors.IsGone(updateErr) || apierrors.IsNotFound(updateErr)) {
-		r.Log.Error(updateErr, "error updating ServiceMeshControlPlane status")
+		r.Manager.GetRecorder(controllerName).Event(r.Instance, "Error", reason, reconciliationMessage)
 	}
+	r.Status.SetCondition(reconciledCondition)
+	r.PostStatus()
```

Comment:
I'm not sure how to get that out, as this is called from `defer` in `Reconcile()`.  FWIW, `Reconcile()` always requeues a request on exit.  It either exits, waiting for a component to become ready, requeues to calculate readiness state after a successful install, or returns an error.

Reply:
@luksa, see the 'defer' statement in 'Reconcile()`.  We're now overriding the `err` return value if there is a problem updating the status and we're not already returning an error.

Reply:
OK, I'm approving this, but be aware that now that errors are returned properly, I am seeing quite a lot of failed status updates due to conflicts: 
```
2019-08-13T20:41:56.470+0200	ERROR	controller_servicemeshcontrolplane	controlplane/reconciler.go:421	error updating ServiceMeshControlPlane status	{"error": "Operation cannot be fulfilled on servicemeshcontrolplanes.maistra.io \"multitenant-install\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/maistra/istio-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/vendor/github.com/go-logr/zapr/zapr.go:128
github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane.(*ControlPlaneReconciler).PostStatus
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane/reconciler.go:421
github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane.(*ControlPlaneReconciler).postReconciliationStatus
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane/reconciler.go:446
github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane.(*ControlPlaneReconciler).Reconcile
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane/reconciler.go:187
github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane.(*ReconcileControlPlane).Reconcile
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/pkg/controller/servicemesh/controlplane/controller.go:269
github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:207
github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157
github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until
	/home/luksa/projects/go/src/github.com/maistra/istio-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
```

Reply:
I thought it was the controller-runtime that was logging these as errors, but it's actually our code. Maybe we can just log them as INFO, since it's a (somewhat) expected event.

Reply:
That's strange.  I don't see any of these when running in the cluster.  I wonder if there's a delay between retrieving the object locally and sending the update request.

Reply:
I take that back.  I'm seeing a bunch of these too. :/

Reply:
And they're all from reconciler.go, line 187.

Reply:
Addressed in #207.